File: 2551_put_marbles_in_bags.java
class Solution {
    public long putMarbles(int[] a, int k) {
        List<Integer> x = new ArrayList<>();
        for (int i = 0; i < a.length - 1; i++) x.add(a[i] + a[i + 1]);
        Collections.sort(x);
        long hi = 0, lo = 0;
        for (int i = 0; i < k - 1; i++) {
            lo += x.get(i);
            hi += x.get(x.size() - 1 - i);
        }
        return hi - lo;
    }
}

File: 2551_put_marbles_in_bags.cpp
class Solution {
public:
    long long putMarbles(vector<int>& w, int k) {
        priority_queue<int> q, p;
        for (int i = 0; i + 1 < w.size(); ++i) {
            int x = w[i] + w[i + 1];
            q.push(x), p.push(-x);
        }
        long long u = 0, v = 0;
        while (--k) u += q.top(), v -= p.top(), q.pop(), p.pop();
        return u - v;
    }
};
